protocols/gossipsub/src/protocol.rs

```diff
@@ -0,0 +1,346 @@
+                    .take_topicIDs()
+                    .into_iter()
+                    .map(TopicHash::from_raw)
+                    .collect(),
+            });
+        }
+
+        let mut rpc_control = rpc.take_control();
+        let mut control_msgs = vec![];
+        // Collect the gossipsub control messages
+        let ihave_msgs: Vec<GossipsubControlAction> = rpc_control
+            .take_ihave()
+            .into_iter()
+            .map(|mut ihave| {
+                GossipsubControlAction::IHave {
+                    topic_hash: TopicHash::from_raw(ihave.take_topicID()),
+                    // TODO: Potentially format the message ids better
+                    message_ids: ihave.take_messageIDs().into_vec(),
+                }
+            })
+            .collect();
+
+        let iwant_msgs: Vec<GossipsubControlAction> = rpc_control
+            .take_iwant()
+            .into_iter()
+            .map(|mut iwant| {
+                GossipsubControlAction::IWant {
+                    // TODO: Potentially format the message ids better
+                    message_ids: iwant.take_messageIDs().into_vec(),
+                }
+            })
+            .collect();
+
+        let graft_msgs: Vec<GossipsubControlAction> = rpc_control
+            .take_graft()
+            .into_iter()
+            .map(|mut graft| GossipsubControlAction::Graft {
+                topic_hash: TopicHash::from_raw(graft.take_topicID()),
+            })
+            .collect();
+
+        let prune_msgs: Vec<GossipsubControlAction> = rpc_control
+            .take_prune()
+            .into_iter()
+            .map(|mut prune| GossipsubControlAction::Prune {
+                topic_hash: TopicHash::from_raw(prune.take_topicID()),
+            })
+            .collect();
+
+        control_msgs.extend(ihave_msgs);
+        control_msgs.extend(iwant_msgs);
+        control_msgs.extend(graft_msgs);
+        control_msgs.extend(prune_msgs);
+
+        Ok(Some(GossipsubRpc {
+            messages,
+            subscriptions: rpc
+                .take_subscriptions()
+                .into_iter()
+                .map(|mut sub| GossipsubSubscription {
+                    action: if sub.get_subscribe() {
+                        GossipsubSubscriptionAction::Subscribe
+                    } else {
+                        GossipsubSubscriptionAction::Unsubscribe
+                    },
+                    topic_hash: TopicHash::from_raw(sub.take_topicid()),
+                })
+                .collect(),
+            control_msgs,
+        }))
+    }
+}
+
+/// A message received by the gossipsub system.
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct GossipsubMessage {
+    /// Id of the peer that published this message.
+    pub source: PeerId,
+
+    /// Content of the message. Its meaning is out of scope of this library.
+    pub data: Vec<u8>,
+
+    /// A random sequence number.
+    pub sequence_number: Vec<u8>,
```

Comment:
In general, for gossipsub the protobuf struct sets this to bytes. So arbitrary implementations can send us bytes for the sequence number. This struct currently represents what we receive from the wire. We then need to interpret the bytes as something. The go-implementation uses a u64, so in the  `id()` function we convert it to a u64 for the `id`, but in general arbitrary other applications could send us larger bytes or arbitrary encodings. Perhaps later this could be generalised from u64.